### PR TITLE
Adding BooleanLiteral type

### PR DESCRIPTION
**Pull request by kozlov-alexey**
This adds new type for representing literal Boolean values supporting
common operators between literal and non-literal arguments.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the patch @kozlov-alexey, I've queued it for review.

File: numba/core/typing/builtins.py [hunk 1]
```diff
@@ -449,7 +445,16 @@ def generic(self, args, kws):
 
 
 @infer_global(operator.ne)
-class ConstOpNotEq(ConstOpEq):
+class ConstOpNotEq(AbstractTemplate):
+    def generic(self, args, kws):
+        assert not kws
+        (arg1, arg2) = args
+        if isinstance(arg1, types.Literal) and isinstance(arg2, types.Literal):
+            return signature(types.boolean, arg1, arg2)
```
Comment (eric-wieser) [inline numba/core/typing/builtins.py:453]: Why doesn't subclassing `ConstOpEq` as was done before work? Seems silly to repeat this function body at any rate.

File: numba/cpython/builtins.py [hunk 2]
```diff
@@ -73,6 +73,20 @@ def opaque_is(context, builder, sig, args):
...
+def bool_is_impl(context, builder, sig, args):
+    arg1, arg2 = args
+    arg1_type, arg2_type = sig.args
+    _arg1 = context.cast(builder, arg1, arg1_type, types.boolean)
+    _arg2 = context.cast(builder, arg2, arg2_type, types.boolean)
+    eq_impl = context.get_function(
+        operator.eq,
+        typing.signature(types.boolean, types.boolean, types.boolean)
+    )
+    return eq_impl(builder, (_arg1, _arg2))
```
Comment (eric-wieser) [inline numba/cpython/builtins.py:86]: Why is this necessary? Doesn't `generic_is` above handle it?

Comment (kozlov-alexey) [reply numba/cpython/builtins.py:86]: @eric-wieser This is to support cases when one arg is BooleanLiteral and other is types.boolean. In this case `generic_is` always returns False since `lhs_type == rhs_type` check is not met.

File: numba/core/typing/builtins.py (hunk 1 above)
Comment (kozlov-alexey) [reply numba/core/typing/builtins.py:453]: @eric-wieser Yep, thanks. I'll revert this change.

File: numba/cpython/builtins.py (hunk 2 above)
Comment (eric-wieser) [reply numba/cpython/builtins.py:86]: Two points would be good to include in comments here:
* This covers both Boolean and BooleanLiteral types
* This is needed for books because bools in python are guaranteed to preserve identity. It's not needed for any other literal types.

File: numba/parfors/parfor_lowering.py [hunk 3]
```diff
@@ -1621,16 +1621,17 @@ def load_potential_tuple_var(x):
                     array_strides.append(strides[j])
                 builder.store(builder.bitcast(ary.data, byte_ptr_t), dst)
         else:
+            aty_is_bool = aty is types.boolean or isinstance(aty, types.BooleanLiteral)
```
Comment (eric-wieser) [inline numba/parfors/parfor_lowering.py:1624]: Shorter and almost as fast as:
```suggestion
            aty_is_bool = isinstance(aty, types.Boolean)
```

Comment (kozlov-alexey) [reply numba/parfors/parfor_lowering.py:1624]: @eric-wieser Yes, I thought about this, but that way it would be less restrictive (any other type that user will subclass from types.Boolean will match and that type might actually use different data model) and with original version we restrict to types actually using BooleanModel that defines 1b width.

Comment (eric-wieser) [reply numba/parfors/parfor_lowering.py:1624]: If you're worried about that, then you should worry about literal subclasses too:

```suggestion
            # do not allow any other subclasses
            aty_is_bool = type(aty) in (types.Boolean, types.BooleanLiteral)
```

File: numba/cpython/builtins.py [hunk 4]
```diff
@@ -84,7 +103,8 @@ def const_eq_impl(context, builder, sig, args):
     return impl_ret_untracked(context, builder, sig.return_type, res)
 
 
-@lower_builtin(operator.ne, types.StringLiteral, types.StringLiteral)
```
Comment (stuartarchibald) [inline numba/cpython/builtins.py:87]: Why is this deleted?

File: numba/parfors/parfor_lowering.py [hunk 5]
```diff
@@ -1621,16 +1621,18 @@ def load_potential_tuple_var(x):
                     array_strides.append(strides[j])
                 builder.store(builder.bitcast(ary.data, byte_ptr_t), dst)
         else:
+            # do not allow any other subclasses
+            aty_is_bool = type(aty) in (types.Boolean, types.BooleanLiteral)
```
Comment (stuartarchibald) [inline numba/parfors/parfor_lowering.py:1625]: `types.BooleanLiteral` inherits from `types.Boolean`, could this just be an `isinstance` check like `aty_is_bool = isinstance(aty, types.Boolean)` ?

Comment (stuartarchibald) [reply numba/parfors/parfor_lowering.py:1625]: I note this is discussed here https://github.com/numba/numba/pull/6015#discussion_r460274711, the current restrictive check does not follow what has been used elsewhere in Numba (or parfors), are there known use cases where this is a problem? Also Numba types have an `is_internal` property to determine if a type is part of the Numba type system or a user defined type, would suggest using that if restrictive behaviour is required.

CC @DrTodd13 any view on the above?

File: numba/cpython/builtins.py (hunk 4 above)
Comment (kozlov-alexey) [reply numba/cpython/builtins.py:87]: This is only to have the same specifications for `eq `and `ne`. Note it's not removed but changed to:
`@lower_builtin(operator.ne, types.Literal, types.Literal)`
which will work for StringLiteral types as well.

Comment (stuartarchibald) [reply numba/cpython/builtins.py:87]: I see, thanks, guess this https://github.com/numba/numba/pull/6015/files#diff-87f8c6b1645b5d35540368237bcc89baR107 is due to https://github.com/numba/numba/pull/6015/files#diff-87f8c6b1645b5d35540368237bcc89baR94? Should probably add a similar note?

File: numba/parfors/parfor_lowering.py (hunk 5 above)
Comment (kozlov-alexey) [reply numba/parfors/parfor_lowering.py:1625]: @stuartarchibald Yes, thank you, using the following definition works too.
`aty_is_bool = isinstance(aty, types.Boolean) and aty.is_internal`

I'm not aware of any issues related to user sub-classing from types.Boolean and affecting parfor lowering, and I agree that changing it in one place only is probably not a good idea. So I can leave it as simple  `isinstance(aty, types.Boolean)` or use above definition. Just need somebody to tell which one to use.

File: numba/cpython/builtins.py (hunk 4 above)
Comment (kozlov-alexey) [reply numba/cpython/builtins.py:87]: Yes, just added it.

File: numba/parfors/parfor_lowering.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/parfors/parfor_lowering.py:1625]: @kozlov-alexey,  @DrTodd13 and I just had an extensive out of band discussion about this and have concluded that a less restrictive `isinstance(aty, types.Boolean)` type check would be good for now. The implications of sub-classing from core Numba types in a breaking manner are not considered elsewhere in the code base and were such a case to occur there's a few options for resolving this, but such a solutions are inevitably quite wide reaching and would impact more than `parfors`. Hows that sound? Thanks!

File: numba/cpython/builtins.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/cpython/builtins.py:87]: Thanks!

File: numba/parfors/parfor_lowering.py (hunk 5 above)
Comment (stuartarchibald) [reply numba/parfors/parfor_lowering.py:1625]: @kozlov-alexey think it's just ^ to resolve then this is good to merge. Thanks.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes, looks good!
